modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/DiscountedCumulativeGainAt.java

```diff
@@ -156,24 +156,29 @@ public EvalQueryQuality evaluate(SearchHit[] hits, List<RatedDocument> ratedDocs
                 unknownDocIds.add(id);
                 if (unknownDocRating != null) {
                     ratings.add(unknownDocRating);
+                } else {
+                    // we add null here so that the later computation knows this position had no rating
+                    ratings.add(null);
                 }
             }
         }
         double dcg = computeDCG(ratings);
 
         if (normalize) {
-            Collections.sort(ratings, Collections.reverseOrder());
+            Collections.sort(ratings, Comparator.nullsLast(Collections.reverseOrder()));
             double idcg = computeDCG(ratings);
             dcg = dcg / idcg;
         }
-        return new EvalQueryQuality(dcg, unknownDocIds);
+        return new EvalQueryQuality(taskId, dcg, unknownDocIds);
     }
 
     private static double computeDCG(List<Integer> ratings) {
         int rank = 1;
         double dcg = 0;
-        for (int rating : ratings) {
-            dcg += (Math.pow(2, rating) - 1) / ((Math.log(rank + 1) / LOG2));
+        for (Integer rating : ratings) {
+            if (rating != null) {
```

Comment:
Correction, can't do this. `computeDCG()` is used twice, first time on an unsorted list where we want to ignore unrated documents but still need to increase the rank counter. For the normalized part you are right, but just for this tiny optimization I don't think we need a separate method or a modifier here.